lsp/nls/src/requests/completion.rs

```diff
@@ -1,16 +1,227 @@
+
+    let name = RE.find(&text)?;
+    Some(name.as_str().chars().rev().collect())
+}
+
+/// Get the identifier before `.<text>` for record completion.
+fn get_identifier_before_field(text: &str) -> Option<String> {
+    // Skip `<text>`
+    let text: String = text
+        .chars()
+        .rev()
+        .skip_while(|c| RE.is_match(c.to_string().as_ref()))
+        .collect::<String>()
+        .chars()
+        .rev()
+        .collect();
+
+    get_identifier_before_dot(&text[..])
+}
+
+fn remove_duplicates(items: &Vec<CompletionItem>) -> Vec<CompletionItem> {
+    let mut seen: Vec<CompletionItem> = Vec::new();
+    for item in items {
+        if !seen.iter().any(|seen_item| seen_item.label == item.label) {
+            seen.push(item.clone())
+        }
+    }
+    seen
+}
+
+/// Search the linearization to find the record information associated with a
+/// partiular ID, and in the scope of a given linearization item.
+fn collect_record_info(
+    linearization: &Completed,
+    item: &LinearizationItem<Types>,
+    name: usize,
+) -> Vec<(Vec<Ident>, Types)> {
+    linearization
+        .get_in_scope(item)
+        .iter()
+        .filter_map(|item| {
+            let (ty, _) = linearization.resolve_item_type_meta(item);
+            match (&item.kind, ty) {
+                // Get record fields from static type info
+                (TermKind::Declaration(..), Types(AbsType::Record(row))) if name == item.id => {
+                    Some((extract_ident(&row), item.ty.clone()))
+                }
+                (TermKind::Declaration(_, _, ValueState::Known(body_id)), _) if name == item.id => {
+                    match find_record_contract_fields(&linearization, *body_id) {
+                        // Get record fields from contract metadata
+                        Some(fields) => Some((fields, item.ty.clone())),
+                        // Get record fields from lexical scoping
+                        None => find_record_fields(&linearization, *body_id)
+                            .map(|idents| (idents, item.ty.clone())),
+                    }
+                }
+                _ => None,
+            }
+        })
+        .collect()
+}
+
+/// Generate possible completion Identifiers given a source text, and its
+/// linearization.
+fn get_completion_identifiers(
+    source: &str,
+    trigger: Option<&str>,
+    linearization: &Completed,
+    item: &LinearizationItem<Types>,
+) -> Result<Vec<CompletionItem>, ResponseError> {
+    let in_scope = match trigger {
+        // Record Completion
+        Some(server::DOT_COMPL_TRIGGER) => {
+            let name = get_identifier_before_dot(source).ok_or(ResponseError {
+                code: ErrorCode::InternalError as i32,
+                message: "Couldn't get identifier for record completion".to_owned(),
+                data: None,
+            })?;
+            let name = Ident::from(name);
+            if let Some(name) = item.env.get(&name).cloned() {
```

Comment:
I find a bit strange to call something `name` which has type `usize`. Especially when it was previously an ident. Wouldn't `id` be a better name for it?